CMake build with find_package support and automatic translation.

alexreinking opened the pull request:
This PR adds CMake build and packaging support... correctly. Users can simply build and install cppfront via:

```
$ cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
$ cmake --build build --target install
```

and then write a simple CMakeLists.txt for their project:

```cmake
cmake_minimum_required(VERSION 3.23)
project(cpp2-example)

find_package(cppfront REQUIRED)
# add_subdirectory(cppfront)
# FetchContent, etc.

add_executable(main main.cpp2)
```

There's a full example in `example/`, too. Users who don't want the cpp2 autodetection can set `CPPFRONT_NO_MAGIC` to `1` before `find_package` and can use `cppfront_enable(TARGETS main)` instead. Alternatively, they can get translated source lists manually via `cppfront_generate_cpp(srcs_var <file.cpp2...>)`.

---

It also augments the test suite to actually build the generated CPP1 files and corrects the tests in response to this. Working on Windows with MSVC and Linux with GCC11. macOS has issues (there's another PR with this). All of this is tested in GitHub Actions workflows.

JohelEGP reviewed (approved):
This is excellent.

> It also augments the test suite to actually build the generated CPP1 files and corrects the headers in response to this.

And it's already paying off.

alexreinking commented:
Trying to rebase this on the latest `main` reveals a syntax error in `cpp2util.h`, here:

https://github.com/hsutter/cppfront/blob/98f6dd46957e100f813245241f183c8aedbcb36f/include/cpp2util.h#L594-L596

The curly braces are not needed.

alexreinking commented:
If there were a good way of picking out which tests had a `main` function and which did not, I would try actually running the binaries (under sanitizers, too).

JohelEGP commented:
You can list them explicitly, as with the `codegen_only_tests` variable. And ensure it's up-to-date by using some tool to inspect the object files for the presence of `::main`.

hsutter commented:
> Trying to rebase this on the latest `main` reveals a syntax error in `cpp2util.h`, here:
> 
> https://github.com/hsutter/cppfront/blob/98f6dd46957e100f813245241f183c8aedbcb36f/include/cpp2util.h#L594-L596
> 
> The curly braces are not needed.

Right, gcc flagged that and I thought I fixed that yesterday after the first commit but now I see I didn't commit that change... done, thanks.

hsutter commented:
> If there were a good way of picking out which tests had a `main` function and which did not, I would try actually running the binaries (under sanitizers, too) and regression-testing their output.

Good point, I was meaning to clean that up -- now all tests have a `main`. Let me know if I missed anything. Thanks!

alexreinking commented:
Looks like it's working. I updated the test harness to build executables, but it's not _running_ them yet. That's the next step.

alexreinking commented:
@hsutter there's a small handful of tests that don't have a corresponding CPP file because they aren't expected to build.

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+)
+
+option(
+    CPPFRONT_INSTALL_RULES "Enable install rules for cppfront" "${PROJECT_IS_TOP_LEVEL}"
```
Any reason why the built-in variable is not used?

https://cmake.org/cmake/help/latest/variable/CMAKE_SKIP_INSTALL_RULES.html

alexreinking replied:
Because of this: https://gitlab.kitware.com/cmake/cmake/-/issues/22561

alexreinking replied:
Oh you mean skipping by checking... yes that is a better idea, thanks.

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+    CPPFRONT_INSTALL_RULES "Enable install rules for cppfront" "${PROJECT_IS_TOP_LEVEL}"
+)
+
+add_executable(cppfront source/cppfront.cpp)
```
```suggestion
add_executable(cppfront_cppfront source/cppfront.cpp)
```

To further recude the chance of possible name collisions, it's good to prefix targets with a project specific name. This prefix can be removed in the install interface with this property:  
https://cmake.org/cmake/help/latest/prop_tgt/EXPORT_NAME.html

alexreinking replied:
I agree with this philosophy _except_ for where it conflicts with the actual project name itself. Who is going to import cppfront and then also name a target cppfront?

JohelEGP replied:
It's probable that there already are CMake buildsystems with a `cppfront` target because cppfront today doesn't provide a CMake buildsystem itself.

alexreinking replied:
I personally think this is paranoid and hyper-defensive, but I'm not going to die on this hill :stuck_out_tongue: 

Done.

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+add_library(cppfront::cpp2util ALIAS cpp2util)
+target_compile_features(cpp2util INTERFACE cxx_std_20)
+target_include_directories(
+    cpp2util INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
```
Since you are going for making this project FC-able, you should make `SYSTEM` conditional, so warnings in cppfront's headers are ignored where supported.

```cmake
set(warning_guard "")
if(NOT PROJECT_IS_TOP_LEVEL)
  set(warning_guard SYSTEM)
endif()

target_include_directories(
    cpp2util ${warning_guard} INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
)
```

JohelEGP replied:
I'm against using `SYSTEM` at all. It goes against "keeping us honest by compiling to C++". It also goes against the valuable trend of keeping/restoring user-facing warnings that occur within the library's implementation.

friendlyanon replied:
`SYSTEM` would apply only when the project is pulled in via `add_subdirectory` (and by proxy FC). This is the default anyway when importing a package in CMake and you have to manually opt-out: https://cmake.org/cmake/help/latest/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html  
This is just about having a consistent interface in all scenarios or at least approaching that as much as possible.

You can also introduce an option to control this behavior in case that is desired, so it can be turned off in all cases:

```cmake
set(warning_guard "")
if(NOT PROJECT_IS_TOP_LEVEL)
  option(
      CPPFRONT_INCLUDES_WITH_SYSTEM
      "Use SYSTEM modifier for cppfront's includes, disabling warnings"
      ON
  )
  mark_as_advanced(CPPFRONT_INCLUDES_WITH_SYSTEM)
  if(CPPFRONT_INCLUDES_WITH_SYSTEM)
    set(warning_guard SYSTEM)
  endif()
endif()
```

JohelEGP replied:
> This is just about having a consistent interface in all scenarios or at least approaching that as much as possible.

So the scenario is `add_subdirectory` without a binary directory, such that cppfront's targets are not imported targets and so `SYSTEM` does not apply by default. Even FC specifies the binary directory. Is there really value in providing a consistent interface for such consumers?

friendlyanon replied:
Binary directory has nothing to do with it. There is no difference between `add_subdirectory(src)` and `add_subdirectory(thirdy-party/foo)`, from CMake's perspective it's all _your_ code that is going to be built together.

`find_package` on the other hand conventionally executes scripts that create `IMPORTED` targets that reference already built artifacts. `IMPORTED` targets have a whole set of implications, such as their headers not belonging to you and warnings from those not being useful, since you can't do anything about them anyway.

JohelEGP replied:
Ah, yes, you're right. I agree it's useful to have for consistency.

JohelEGP replied:
> such as their headers not belonging to you and warnings from those not being useful, since you can't do anything about them anyway.

When that's really the case, the library should workaround that. For example, see https://github.com/mpusz/units/issues/203. That was inspired by some r/cpp posts about MS's STL restoring user-facing warnings. I'd have linked them then, but I lost their track.

JohelEGP replied:
That's tangential to what I'm referring to. `SYSTEM` behavior prevents `std::duration<float>(1.0)` from warning on the conversion from `double` to `float`, if your implementation effectively does `float(1.0)`, as opposed to the mandated `static_cast<float>` which MS's STL moved away from in some places.

JohelEGP replied:
To be clear, I want your second suggested solution which allows disabling `SYSTEM` behavior: https://github.com/hsutter/cppfront/pull/15#discussion_r981142370.

While I appreciate the consistency your suggestion brings, it should be better. The status-quo is for everyone to do it their own way, probably inconsistently, if they do it at all. I wish this simply were one of CMake's `<PROJECT-NAME>` [Variables that Change Behavior](https://cmake.org/cmake/help/latest/manual/cmake-variables.7.html#id3). Isn't it in the purview of CMake to facilitate bridging the gap between `find_package` and `add_subdirectory`?

Until then, I'd appreciate having a (function in a) module which defines these for me.
```CMake
include(JEGPBridgeFindPackageAndAddSubdirectory)
# ...
if(JEGP_cppfront_INCLUDES_WITH_SYSTEM) # Probably unnecessary for a built-in.
# ...
```

alexreinking replied:
@friendlyanon @JohelEGP - I implemented this with an option called `CPPFRONT_NO_SYSTEM`.

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+    include(GNUInstallDirs)
+    include(CMakePackageConfigHelpers)
+
+    set(CPPFRONT_INSTALL_CMAKEDIR "${CMAKE_INSTALL_DATADIR}/cmake/cppfront"
+        CACHE STRING "Install location for cppfront's platform-independent CMake files")
+
+    set(CPPFRONT_INSTALL_EXE_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/cppfront-exe"
+        CACHE STRING "Install location for cppfront's platform-specific CMake files")
```
I'd mark these advanced personally. The defaults are what one would expect, so fiddling with them is not really expected.

alexreinking replied:
Sure

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,127 @@
+        COMPONENT cppfront
+    )
+
+    configure_package_config_file(
+        cmake/cppfront-exe-config.cmake.in
+        cmake/cppfront-exe-config.cmake
+        INSTALL_DESTINATION "${CPPFRONT_INSTALL_EXE_CMAKEDIR}"
+        PATH_VARS CMAKE_INSTALL_BINDIR
+    )
```
I'd avoid using `configure_package_config_file` and just have a simple config file instead. The config file doesn't even use the `PACKAGE_CMAKE_INSTALL_BINDIR` that is set by `PATH_VARS`. The `INSTALL_DESTINATION` variable also embeds an install prefix that is decided at configure time, but since CMake 3.15, that is at odds with `cmake --install ... --prefix ...`. The config scripts needn't be configureable imo.

alexreinking replied:
> The config file doesn't even use the `PACKAGE_CMAKE_INSTALL_BINDIR` that is set by `PATH_VARS`.

Yes, it does:

https://github.com/hsutter/cppfront/blob/84982607c2901d9aecb26ef03534136f33337865/cmake/cppfront-exe-config.cmake.in#L1-L7

> The `INSTALL_DESTINATION` variable also embeds an install prefix that is decided at configure time

No, it doesn't. It's used to compute a relative path to _whichever_ prefix is selected down the line. Here is the generated code, so you can see what it actually does:

```
$ cmake -S . -B build -DCMAKE_INSTALL_PREFIX=$PWD/foobar
$ cmake --build build --target install
$ cat foobar/lib/cmake/cppfront-exe/cppfront-exe-config.cmake 
# ...
get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../../../" ABSOLUTE)
# ...
include("${CMAKE_CURRENT_LIST_DIR}/cppfront-exe-targets.cmake")

set_and_check(CPPFRONT_EXECUTABLE "${PACKAGE_PREFIX_DIR}/bin/cppfront")
check_required_components(cppfront-exe)
```
As you can see, this does not mention `foobar` at all, and the package is perfectly relocatable.

friendlyanon replied:
Okay then, I'm just blind and didn't see the variable used.

After a closer inspection of the module's docs, I can certainly say that I totally mixed `INSTALL_PREFIX` and `INSTALL_DESTINATION` up. I also found that `INSTALL_DESTINATION` is not explained in greater detail. [Two sentences](https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html#:~:text=path%3E%20given%20to-,INSTALL_DESTINATION,-must%20be%20the) and they don't say much :(

friendlyanon replied:
Also, `CMAKE_EXECUTABLE_SUFFIX` appears to have a `<LANG>` variant, now that I'm looking at your example.

friendlyanon commented on cmake/cppfront-config.cmake.in:
```diff
@@ -0,0 +1,10 @@
+include("${CMAKE_CURRENT_LIST_DIR}/cppfront-targets.cmake")
+include("${CMAKE_CURRENT_LIST_DIR}/CppfrontHelpers.cmake")
+
+check_required_components(cppfront)
```
This package has no _package components_, this check is unnecessary. I think you are confusing them with _install components_.

_Install components_ as defined by `install(... COMPONENT ...)` are used by CPack and `cmake --install ... --component ...` only.

_Package components_ are entirely dealt with in the config script. Here is an [example](https://github.com/friendlyanon/SFML/blob/feature/cmake-rework/cmake/SFMLConfig.cmake.in#L23).

alexreinking replied:
No, I'm not. That macro will throw an error if a component is incorrectly specified. Ie. `find_package(cppfront REQUIRED foobardoesnotexist)`.

JohelEGP replied:
> This macro should be called even if the package doesn't provide any components to make sure users are not specifying components erroneously.
> -- https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html

friendlyanon replied:
I find it funny that you find prefixing target names hyper-defensive, but this isn't that :)

I guess you're right, this is stricter and doesn't lose anyone anything. "I'm not going to die on this hill" 😀

alexreinking replied:
> I find it funny that you find prefixing target names hyper-defensive, but this isn't that :)

@friendlyanon - yes, we all have our own value functions for such things. I won't pretend mine is perfect. :slightly_smiling_face:

friendlyanon commented on cmake/CppfrontHelpers.cmake:
```diff
@@ -0,0 +1,86 @@
+    cmake_parse_arguments(PARSE_ARGV 0 ARG "" "" "TARGETS")
+
+    foreach (tgt IN LISTS ARG_TARGETS)
+        get_property(sources TARGET "${tgt}" PROPERTY SOURCES)
+        list(FILTER sources INCLUDE REGEX "\\.cpp2$")
```
While pretty clever, I'm not sure relying on CMake ignoring `cpp2` files right now may work out in the future. I think doing the same thing that the CMake provided Protobuf module does would be more future proof.

alexreinking replied:
I could re-work it to do both. I like the end-user simplicity of this solution, though, and if CMake upstream ever gains support for `.cpp2` files, that will (a) be guarded by `enable_language` one way or another and (b) utterly supersede this code.

alexreinking replied:
I added `cppfront_generate_cpp`, which is called internally by `cppfront_enable(TARGETS)` (which itself is called by the "magic" target scanning)

friendlyanon commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,46 @@
+      - name: Build regression tests
+        run: cmake --build build/regression-tests --config Release
+      - name: Run regression tests
+        run: ctest --output-on-failure -C Release -j 4
```
Why 4 jobs? Both the Windows and Linux runners have only 2 cores: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

alexreinking replied:
Two is better, then, good point.

friendlyanon commented on CMakeLists.txt:
```diff
@@ -0,0 +1,153 @@
+target_sources(
+    cppfront_cppfront
+    PRIVATE
+    FILE_SET HEADERS
```
Speaking of prefixes, are file sets global or directory scoped?

JohelEGP replied:
Target scoped:
> Targets have zero or more named file sets.
> [...]
> The following target properties are set by target_sources(FILE_SET), but they should not generally be manipulated directly:
> -- https://cmake.org/cmake/help/latest/command/target_sources.html#file-sets

friendlyanon reviewed (approved):
I haven't tested the code in `CppfrontHelpers.cmake` locally, but looks like your fork's CI is green.

Besides the lack of documentation, beautiful PR!

hsutter commented:
Interim ack: Thanks! I'm potentially convinceable but I'm really trying hard not to complicate things with CMake right now. I fully admit that could be because of my innate aversion to dipping even a toenail into the build system wars (at least yet, at least if I can avoid it). Perhaps for now this could be a separate sister project I link to from the wiki (which I'm about to set up soon)?

> there's a small handful of tests that don't have a corresponding CPP file because they aren't expected to build.

> Also, can I ask why the cpp2 files are duplicated in test-results?

No particular reason, it was just simpler to have the parent contain the tests and then copy everything into the working directory. I'm about to push a commit that changes that to remove the duplication (and changes the filename extension for the test output results) as part of getting the regression tests to also execute and to work around Windows short filename legacy weirdness (a gift that keeps on giving in 2022, durnit). I do want to actually run the tests, so I can spot regression errors in `cpp2util.h` -- thanks for reporting a few already.

More soon, probably this weekend or next weekend (I'm mainly working on this on nights/weekends).

alexreinking commented:
> Interim ack: Thanks!

Thanks for your time looking at this PR!

> I'm really trying hard not to complicate things with CMake right now.

I totally understand. My fallback goal with this PR (if it can't be merged) is to at least _show_ what it takes to do CMake right for this project. I think the other attempts so far have fallen far short.

> Perhaps for now this could be a separate sister project I link to from the wiki (which I'm about to set up soon)?

I can, of course, re-work this into a sister project which includes cppfront as a git submodule. If we go that route, I do hope that we can work together to get the regression tests in this repository into a state such that I don't have to touch the test harness every time I bump the submodule pointer.

> I'm potentially convinceable [...] I do want to actually run the tests, so I can spot regression errors in cpp2util.h -- thanks for reporting a few already.

You're welcome!

Perhaps the test harness I built will overcome other reservations you might have. [It already works](https://github.com/alexreinking/cppfront/actions/runs/3138721950/jobs/5098368003) on both Linux and Windows and can be easily enabled on macOS, once outstanding bugs there are fixed. 

If you're worried about ongoing maintenance, I can offer my time addressing issues related to CMake. I can say this with confidence since I will have to do the same things maintaining a sister project, anyway. I have very relevant experience as a maintainer of [Halide](https://github.com/halide/Halide).

> I fully admit that could be because of my innate aversion to dipping even a toenail into the build system wars (at least yet, at least if I can avoid it).

Still, it's worth noting that [CMake has been the most popular](https://www.jetbrains.com/lp/devecosystem-2019/cpp/) project model and build system for C++ since 2019. In the last year, even [Bryce Adelstein Lelbach has taken to calling CMake](https://www.youtube.com/watch?t=4994&v=OgM0MYb4DqE) the _de facto_ standard build system, noting that it now has majority market share.

For better or worse, if one must choose a single, portable, build system for C++ (and first-party support for multiple is, imho, unsustainable) it should pretty clearly be CMake.

hsutter commented:
Thanks @alexreinking, yes please let's set it up as a sister project for now, and that will help me especially in the short term to catch up on the issues being added here and in the medium term implementing the rest of the basic language. Then it will be a great time to look at merging. Thanks again for the interest and expertise, and for understanding!

Just let me know here in this issue when you have a link to the project, and I'll add it to the cppfront wiki home page. I'll keep this issue open until then to help me remember.

alexreinking commented:
@hsutter - thanks! That all sounds like a good plan to me.

Here is the link I'd like to use: https://github.com/modern-cmake/cppfront

You can describe it as:

> A modern CMake (3.23+) build for cppfront, complete with automatic cpp2-to-cpp1 translation. Compatible with `find_package`, `add_subdirectory`, and FetchContent.

---

I wonder if I could also ask you to please add the missing headers in `mixed-captures-in-expressions-and-postconditions`:

```
#include <algorithm>
#include <vector>
```

alexreinking commented:
@hsutter - awesome, thanks!
